strongbox-security/strongbox-authentication-support/src/main/java/org/carlspring/strongbox/authentication/support/AuthoritiesExternalToInternalMapper.java

```diff
@@ -59,8 +74,35 @@ public void afterPropertiesSet()
         Assert.notEmpty(rolesMapping, "rolesMapping property not set");
     }
 
+    /**
+     * @return Returns an immutable map containing the same entries as {@link #rolesMapping}
+     */
+    public Map<String, String> getRolesMapping()
+    {
+        return ImmutableMap.copyOf(rolesMapping);
+    }
+
     public void setRolesMapping(Map<String, String> rolesMapping)
     {
-        this.rolesMapping = rolesMapping;
+        Assert.notNull(rolesMapping, "rolesMapping cannot be null");
+        this.rolesMapping = new ConcurrentHashMap<>(rolesMapping);
+    }
+
+    public String upsertRoleMapping(String externalRole,
```

Comment:
Upsert means insert if doesn't exist, update otherwise :) I'll change to "put" :)